Implement query pagination

chriszarate opened the pull request:
Based on #337. ~Still working on docs, but feel free to review!~ Ready for review!

chriszarate committed:
Update Art Institute example

chriszarate committed:
Update documentation for pagination

chriszarate committed:
Implement pagination

brookewp commented:
This is an exciting addition! 🎉 

I tested out the pagination page selection dropdown after searching and the data never loaded or timed out (maybe eventually it does but I couldn't wait forever):

https://github.com/user-attachments/assets/29133042-8107-4bd6-bb40-8a192c8d1933

I didn't run into this using the pagination dropdown on the initial un-searched set or when searching and manually clicking the next page button.

chriszarate commented:
@brookewp Thanks! FYI inspecting the network response is useful here. Above page 90 or so I get an error from the API:

> Client error: `GET https://api.artic.edu/api/v1/artworks/search?q=van+gogh&limit=12&page=90` resulted in a `403 Forbidden` response:\n{\"status\":403,\"error\":\"Invalid number of results\",\"detail\":\"You have requested too many results. Please refine your para...

It seems like the reported total items for this search (`2341`) does not align with reality, since `90 * 12 = 1,080`. 

It does point to the need to catch and surface these errors. What would be the best way to do this in DataViews / the block editor?

brookewp commented:
>It does point to the need to catch and surface these errors. What would be the best way to do this in DataViews / the block editor?

I don't think there is any way to do this in `DataViews` currently... 🤔 the only control we have is over the data, if there is no data, (an empty array passed to the `data` prop), it shows 'No results found' without the table, but then we lose the pagination:  

![Storybook example](https://github.com/user-attachments/assets/e4d8128b-27cf-48df-9400-ed1d420aafe9)

Storybook example here: [Storybook example link](https://wordpress.github.io/gutenberg/?path=/story/dataviews-dataviews--empty)

The only alternative I can think of right now is: 
1. Hide the page dropdown via CSS (I don't think it adds much value, as I don't think anyone would know which page contains which data. I can only think of consumers going to the end, but if we add filtering, they could reverse the order alphabetically to achieve the same result). 
2. Replace the table element via a portal 😬 to replace with our own error screen. 

What do you think?

ingeniumed commented:
Using [error notices](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-notices/) could be a potential solution to this. It can either be a snackbar in a corner, or a banner at the very top so as to allow the editor itself to control it. I have been trying to experiment with this in dataviews and it's been tricky to get it working as the notice shows up on the editor while the dataview is covering it up. You also get a warning saying that you can't render a notice when the itemList is still being rendered. I suspect that's due to me putting the errorNotice within `itemList.tsx`. 

There is a prototype that @oandregal has made [here](https://github.com/wptrainingteam/devblog-dataviews-plugin/blob/main/src/App.js#L152) that might shed some light on how to use it correctly

chriszarate committed:
Merge remote-tracking branch 'origin/trunk' into add/query-pagination

chriszarate committed:
Update Query docs

chriszarate committed:
Fix a few schema references

chriszarate committed:
Merge branch 'improve/declarative-search' into add/query-pagination

brookewp commented:
Data doesn't show for any of my **Airtable** sources. However, I can select an item and then see it in pattern selection.

| Trunk  |  PR |
| ------------- | ------------- |
|  <img width="900" alt="Screenshot 2025-02-04 at 3 51 22 PM" src="https://github.com/user-attachments/assets/2c944d2c-9e8f-4b4c-8a2a-ff9bdd69d448" />  | <img width="1000" alt="Screenshot 2025-02-04 at 3 45 12 PM" src="https://github.com/user-attachments/assets/b4df5f44-7499-44e8-a4bc-0926d14e1ac5" /> |

**Shopify** data loads, but no pagination and 'Item Price' column is missing (but shows when selecting like above): 

| Trunk  |  PR |
| ------------- | ------------- |
|  <img width="900" alt="Screenshot 2025-02-04 at 3 49 42 PM" src="https://github.com/user-attachments/assets/d87b3f14-13d0-4404-9cfd-8596d5012cd3" /> | <img width="1000" alt="Screenshot 2025-02-04 at 3 46 12 PM" src="https://github.com/user-attachments/assets/810bb281-a656-4fa1-937e-19f81e44c6bb" />  |

But of course the example I tested in my earlier comment still works 😕 : 

<img width="1097" alt="Screenshot 2025-02-04 at 3 48 40 PM" src="https://github.com/user-attachments/assets/2f7466a4-ef1e-4012-9f2a-eaaebdd5ab23" />

chriszarate committed:
Move field filter to useful place

chriszarate committed:
Move removeNullValues to util

chriszarate committed:
Refactor ItemList to remove memos

chriszarate commented:
> Data doesn't show for any of my **Airtable** sources. 

Thanks for testing! I've had to fix some pretty gnarly merge conflicts and I flubbed one. Fixed in https://github.com/Automattic/remote-data-blocks/pull/338/commits/0613bf3527dbc3f4cf119c86c070c090eedd442b.

That said, the logic with all the memos and the effect is really hard to follow, so I cleaned it up a bit in https://github.com/Automattic/remote-data-blocks/pull/338/commits/33f04d09757d554f6811c8afc390293afd7ac144 and https://github.com/Automattic/remote-data-blocks/pull/338/commits/eed54e2dedff4db8b5610738dbfec724f7605f56. It makes the diff a bit more severe but hopefully the resulting code is easy to follow.

maxschmeling commented on src/blocks/remote-data-container/components/item-list/ItemList.tsx:
```diff
@@ -1,135 +1,133 @@
+					: undefined,
+			enableSorting: field !== mediaField,
+		};
+	} );
```
```suggestion
	const fields = fieldNames.map( field => ( {
		id: field,
		label: availableBindings[ field ]?.name ?? field,
		enableGlobalSearch: true,
		getValue: ( { item }: { item: RemoteDataResult } ) => item[ field ] as string,
		render:
			field === mediaField
				? ( { item }: { item: RemoteDataResult } ) => {
						return (
							<img alt={ ( item.image_alt as string ) ?? '' } src={ item[ field ] as string } />
						);
				  }
				: undefined,
		enableSorting: field !== mediaField,
	} ) );
```

brookewp commented on src/blocks/remote-data-container/components/item-list/ItemList.tsx:
```diff
@@ -1,135 +1,133 @@
-		}
-	}, [ view, searchTerms ] );
-
-	// filter, sort and paginate data
-	const { data: filteredData, paginationInfo } = useMemo( () => {
-		return filterSortAndPaginate( data ?? [], view, fields );
-	}, [ data, view ] );
```
@chriszarate are we removing the search for upcoming changes in #337? 

I noticed because our integrated sources don't have pagination, since now it'll fallback to `data.length` if pagination isn't explicitly set: https://github.com/Automattic/remote-data-blocks/blob/eed54e2dedff4db8b5610738dbfec724f7605f56/src/blocks/remote-data-container/components/item-list/ItemList.tsx#L107-L108
https://github.com/Automattic/remote-data-blocks/blob/eed54e2dedff4db8b5610738dbfec724f7605f56/src/blocks/remote-data-container/components/item-list/ItemList.tsx#L154-L157

chriszarate replied:
@brookewp The reasons for removing search (and filtering) are summarized here:

https://wp.me/pfvsEV-tn#behavior-change-for-searching-sorting-and-filtering-in-dataviews

After this PR is approved, I'll work on implementing pagination for our other integrations.

maxschmeling reviewed (approved):
The issues we were seeing are fixed. This looks great. I don't see anything major, though it'll be nice to get this in and see how it works as we use it more.

chriszarate committed:
Tighten up

Co-authored-by: Max Schmeling <max.schmeling@automattic.com>

chriszarate committed:
Merge remote-tracking branch 'origin/improve/declarative-search' into add/query-pagination

brookewp reviewed (approved):
Data is loading again 🎉 Looks good to me!
